### PR TITLE
Changing billing cycle no longer requires cancelling first

### PR DESCRIPTION
**Pull request by brianium**

### CODE REVIEW

File: src/Processor/Braintree/Subscription/Update/ChangeBillingCycleStrategy.php [hunk 1]
```diff
@@ -64,12 +68,36 @@ private function createReplacementSubscription(Subscription $original, Subscript
...
+
+        if (0.0 === $fuzzy) {
+            return null;
+        }
+
+        if ($fuzzy >= 0.0) {
+            return new Discount('balance', 1, new Price($fuzzy));
+        }
+
+        return new AddOn('overdue', 1, new Price(abs($fuzzy)));
```
Comment (brianium) [inline src/Processor/Braintree/Subscription/Update/ChangeBillingCycleStrategy.php:98]: @johncorrelli we don't have a test case for this in the suite, but we never did. Can you spot check this for correctness?

File: src/Processor/Braintree/Mapper/SubscriptionMapper.php [hunk 2]
```diff
@@ -136,10 +136,30 @@ public function fromResult(Braintree\Subscription $result): Subscription
...
+        $cycle = $billingPeriod->getBillingCycle();
+        $remainingCycle = $billingPeriod->getRemainingBillingCycle();
+
+        $costPerDay = $cost / $cycle;
+        $remainingValue = $costPerDay * $remainingCycle;
+
+        $remainingValue = new Money(round($remainingValue, 2));
 
-        return $subscription->setInitialDiscounts(new Modifiers($discounts));
+        return $sub->setRemainingValue($remainingValue);
```
Comment (johncorrelli) [inline src/Processor/Braintree/Mapper/SubscriptionMapper.php:162]: should we avoid the assignment/reassignment and do....

```php
 $remainingValue = round($costPerDay * $remainingCycle, 2);

return $sub->setRemainingValue(new Money($remainingValue));
```

**Review by johncorrelli — CHANGES_REQUESTED** [review]
Nice work homie! Just a few little things. Most importantly around a subscription without a price and the function naming on the billing period.

File: src/Model/Subscription/BillingPeriod.php [hunk 3]
```diff
@@ -25,4 +25,16 @@ public function getStartDate(): DateTime
     {
         return $this->startDate;
     }
+
+    public function getBillingCycle(): int
```
[the remark this answers is not in the document]
Comment (brianium) [reply src/Model/Subscription/BillingPeriod.php:29]: Braintree just called it cycle. A doc block sounds good to me

File: src/Model/Modifier/Modifier.php [hunk 4]
```diff
@@ -32,6 +34,17 @@ public function setPrice(Price $price): Modifier
         return $this;
     }
 
+    public function getValue(): Money
+    {
+        $price = $this->price;
+
+        if (null == $price) {
```
[the remark this answers is not in the document]
Comment (brianium) [reply src/Model/Modifier/Modifier.php:41]: Guess not! I can change it though 🤷🏻

File: src/Model/Subscription.php [hunk 5]
```diff
@@ -155,6 +174,26 @@ public function getBalance(): ?Money
         return $this->balance;
     }
 
+    public function getValue(): Money
+    {
+        $price = $this->price;
+
+        if (null === $price) {
+            return new Money(0.0);
+        }
```
[the remark this answers is not in the document]
Comment (brianium) [reply src/Model/Subscription.php]: @johncorrelli please check out what I did instead with regards to this.

**Review by johncorrelli — APPROVED** [review]
:boom: Looks great man! And lovin' the automated tests we have in place now! :heart:

File: src/Model/Subscription.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (johncorrelli) [reply src/Model/Subscription.php]: Looks good to me! Now you can have a subscription without a price, but the modifiers can drive the actual amount that is being charged. :heart:

**Pull request merged**